packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -1383,9 +1387,29 @@ If you are using a wallet or dapp, try resetting your wallet's accounts.`
         // run the call in a block that is as close to the real one as
         // possible, hence putting the tx to the block is good to have here.
         await this._addTransactionToBlock(blockContext, tx);
+
+        if (workaroundEthCallGasLimitIssue) {
+          const txGasLimit = new BN(tx.gasLimit);
+          const blockGasLimit = new BN(blockContext.header.gasLimit);
+
+          if (txGasLimit.gt(blockGasLimit)) {
+            blockContext.header.gasLimit = tx.gasLimit;
+          }
+        }
       } else {
-        // if the context is to run calls with the latest block
-        blockContext = await this.getLatestBlock();
+        // if the context is to run calls with a block
+        // We know that this block number exists.
```

Comment:
How do we know that the block number exists?

Reply:
I'll add a better explanation.